### ISSUE TITLE
Bug: [no-unnecessary-condition] Optional chaining on array false positive

### ISSUE THREAD

**Issue by jakeleventhal** [typescript-eslint/typescript-eslint#7505]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.1-rc&sourceType=script&fileType=.ts&code=MYewdgzgLgBA5gUzAQQE6oIYE8YF4YAUAlAFwwDeMGZA5AG4YA2AjDTAD4z1MBMbAvgG0AungB8FAFAwYqBFACuqMDBEBuSf0mTQkWBnR54SNJizENuiCEYIAdIxBwCB1IIAMwgPx2MHTjQYEAAmAGY0REA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=true

### Repro Code

```typescript
See link

It is wrong that the optional chaining is not necessary
```

### ESLint Config

```javascript
see link
```

### tsconfig

```jsonc
see link
```

### Expected Result

No error (assume the func returns any length of array.

### Actual Result

see link

**Comment by bradzacher** [comment]
In general we recommend turning on [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess) because TS encodes this sort of thing explicitly in the types - allowing you to ensure your codebase is safe everywhere.

As this encodes the undefined possibility in the types it also powers the rule to correctly ensure you don't have unnecessary conditions.

We do have some very limited handling in the rule to ignore array index access - but ofc that does help you be safe in the first place.
